### ISSUE TITLE
Version 2 wishlist

### ISSUE THREAD

**Issue by DarkoKukovec** [infinum/datx#110]
Since version 1 is almost done, and it took a little bit longer than expected, let's start a wishlist for version 2 - features that would be nice to have, but would be a breaking change:

* [x] Remove `CompatModel` & `CompatCollection`
* [x] Remove deprecated `find` and `remove`
* [x] Remove `setupModel`
* [x] JSON API - disallow numeric IDs
* [x] Rename `params` to `queryParams.custom`
* [x] Separate `IRequestOptions` into three parts - `queryParams`, `networkConfig`, `cacheOptions`
* [x] #132 - Implement "Buckets" [INTERNAL]
* [x] #103 - Polymorphic relationships
* [x] #131 - Caching revamp
* [x] #34 - `fetchAll` revamp
* [x] [Write a migration guide, update docs](https://datx.dev/docs/next/migration-guide/from-v1)
* [x] #172 - consistent response data
* [x] #173 - endpoint fn with explicit baseurl
* [x] #154 - Prop mapping
* [x] #164 - Polymorphic dynamic model types
* [x] #130 - Add model dirty states
----------
* [x] #133 - Query builder
* [ ] #153 - Read-only props

# Breaking changes
This list is made based on the current state of the `release/v2` branch

## `id` and `type` fields
In v1, those weren't consistent - in some cases `id` and `type` were used to define the model, but in some other cases they had to be defined separately. In v2, this will be consistent and it will need to be defined explicitly.

## Removed deprecated stuff
`CompatCollection`, `CompatModel`, `collection.find`, `collection.remove`, `setupModel`

## Stricter action support
In v1, some of the mutations have been automatically wrapped into actions. In v2 this is only done in some cases where multiple changes are being done (e.g. the `model.update()` method). This will be a breaking change if you're using the MobX strict mode

## Snapshot/patch format changes
The serialised format of the model is a bit different, more specifically the meta object inside of it.

## References to missing models
In v1, if a model had a reference to another model that is not in a collection, it would return the model id instead of the instance. In v2, such models will be skipped - this means that the reference value will always be either a model or an array of models.

## JSON API pagination
The pagination is now using methods instead of just plain property access: `await response.next` becomes `await response.next()`

## Caching
`config.cache` is no longer a boolean, but a `CachingStrategy` enum instead. `true` maps to `CachingStrategy.CACHE_FIRST` and `false` maps to `CachingStrategy.NETWORK_ONLY`.

## JSON API Collection caching
The static `cache` property on the collection is no longer a boolean. It's a `CachingStrategy` enum instead, without any default value (the global fallback is used)

## Response data is `null` instead of `[]` when no data
In v1, empty data would fall back to `[]` during response parsing, which is not technically correct. In v2, the response data value will be the same as the API response.

## Endpoint function base URL
In v1, the return value of the endpoint function would be appended to the base url, which made some operations where a model has a different base url much harder. In v2, this is not done (it's still done when there is no endpoint or when it's a string) - instead, the endpoint function receives a baseurl parameter, and you are responsible of prepending it to your endpoint url.

# Deprecations

## `@prop` attribute
* `@prop` -> `@Attribute()`
* `@prop.toMany(Foo)` -> `@Attribute({ toMany: Foo })`
* etc.

## `fetch` and `fetchAll`
Deprecated in favour of `getOne` and `getMany`. They use the new methods internally, but force caching strategies that were used in the old methods before, to keep the compatibility.

# Major new stuff

## JSON API getOne/getMany
`getOne` and `getMany` behave similarly to `fetch` and `fetchAll` but make use of better caching strategies.
Caching options can be defined globally (config object), on the collection level and on the request level. The options contain a maxAge value in seconds and the caching strategy type. Available caching strategies are:
```typescript
export enum CachingStrategy {
  NETWORK_ONLY, // Ignore cache
  NETWORK_FIRST, // Fallback to cache only on network error
  STALE_WHILE_REVALIDATE, // Use cache and update it in background
  CACHE_ONLY, // Fail if nothing in cache
  CACHE_FIRST, // Use cache if available
  STALE_AND_UPDATE, // Use cache and update response once network is complete
}
```

## JSON API Collection caching
The collection has a new static `maxCacheAge` value that is the maximal age of the cache in seconds. The default value is `undefined` (fallback to the global value).
The collection `toJSON` will now also serialise and hydrate the cache that was linked to that collection.

**Comment by svobik7** [comment]
Hi there, v2 roadmap looks pretty good 👍thank you for such a great lib which datx is 🙂 is there any time est on reaching v2 stable?

**Comment by DarkoKukovec** [comment]
Thanks @svobik7 
There is no exact estimate yet, but the plan is to do all the breaking changes first, then release 2.0 as early as possible and then add other features that are not breaking (like #153 and #154) in next 2.x releases
